core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/datastore/request/version/CodeSystemVersionCreateRequest.java

```diff
@@ -124,18 +124,58 @@ public Boolean execute(ServiceProvider context) {
 		} catch (NotFoundException e) {
 			// ignore
 		}
+
+		final Map<CodeSystemEntry, IVersioningManager> versioningManagersByCodeSystem = createVersioningManagers(context, codeSystem);
+		for (Entry<CodeSystemEntry, IVersioningManager> entry : versioningManagersByCodeSystem.entrySet()) {
+			createVersion(context, entry.getValue(), entry.getKey(), user);
+		}
 		
-		// check that the specified effective time
-		validateEffectiveTime(context, codeSystem);
+		return Boolean.TRUE;
+	}
+	
+	private Map<CodeSystemEntry, IVersioningManager> createVersioningManagers(final ServiceProvider context, final CodeSystemEntry codeSystem) {
+		final Map<CodeSystemEntry, IVersioningManager> versioningManagersByCodeSystem  = Maps.newHashMap();
+		
+		final IVersioningManager versioningManager = VersioningManagerBroker.INSTANCE.createVersioningManager(codeSystem.getTerminologyComponentId());
+		versioningManagersByCodeSystem.put(codeSystem, versioningManager);
+
+		if (versioningManager.needsOtherCodeSystemsToVersion()) {
+			final Map<String, String> codeSystemsByRepositoryId = versioningManager.codeSystemDependenciesByRepositoryId();
+			for (Entry<String, String> entry : codeSystemsByRepositoryId.entrySet()) {
+
+				CodeSystemEntry cs = null; 
+				try {
+					cs = fetchCodeSystem(context, entry.getKey(), entry.getValue());
+				} catch(NotFoundException e) {
+					throw new CodeSystemNotFoundException(entry.getValue());
+				}
+				
+				final IVersioningManager otherVersioningManager = VersioningManagerBroker.INSTANCE.createVersioningManager(cs.getTerminologyComponentId());
+				versioningManagersByCodeSystem.put(cs, otherVersioningManager);
+			}
+		}
 		
+		return versioningManagersByCodeSystem;
+	}
+	
+	
+	private CodeSystemEntry fetchCodeSystem(ServiceProvider context, final String repositoryId, final String codeSystemShortName) {
+		return CodeSystemRequests.prepareGetCodeSystem(codeSystemShortName)
+			.build(repositoryId)
+			.execute(context.service(IEventBus.class))
+			.getSync();
+	}
+
+	private void createVersion(ServiceProvider context, IVersioningManager versioningManager, CodeSystemEntry codeSystem, String user) {
```

Comment:
Reusing this method in a `forEach` to version all dependent code systems as well long with the current one certainly does its job, but we can improve it to limit the number of unwanted content changes while the versioning runs for a single code system.

Refactor the versioning process to `acquire` and `release` locks for all target code systems at once at the beginning and at the end of the process.